OnBaseDocsApi/Models/ProfileCollection.cs

```diff
@@ -54,61 +70,62 @@ public Application LogIn(string profileName)
 
         public void Refresh()
         {
-            // Make a copy of the existing credentials.
-            Dictionary<string, Profile> newProfiles;
-            lock (Lock)
-            {
-                newProfiles = Profiles.ToDictionary(
-                    x => x.Key,
-                    x => new Profile
-                    {
-                        Application = null,
-                        Credential = x.Value.Credential,
-                    });
-            }
-
-            // Login each credential so that we get a new OnBase Application.
-            LogInAll(newProfiles);
-
-            // Swap in the new profiles.
-            Dictionary<string, Profile> oldProfiles;
-            lock (Lock)
-            {
-                oldProfiles = Profiles;
-                Profiles = newProfiles;
-            }
+            LogInAll();
+        }
 
-            // Log out all of old OnBase Application.
-            LogOutAll(oldProfiles);
+        Application LogIn(ApiConfig config, Credential cred)
+        {
+            var props = Application.CreateOnBaseAuthenticationProperties(
+                config.ServiceUrl,
+                cred.Username,
+                cred.Password,
+                config.DataSource
+            );
+            return Application.Connect(props);
         }
 
-        static void LogInAll(Dictionary<string, Profile> profiles)
+        void LogInAll()
         {
             var config = Global.Config;
 
-            Task.WaitAll(profiles.Select(profile => Task.Run(() =>
+            Task.WaitAll(Profiles.Select(p => Task.Run(() =>
             {
                 try
                 {
-                    var props = Application.CreateOnBaseAuthenticationProperties(
-                        config.ServiceUrl,
-                        profile.Value.Credential.Username,
-                        profile.Value.Credential.Password,
-                        config.DataSource
-                    );
-                    profile.Value.Application = Application.Connect(props);
+                    var currApp = p.Value.Application;
+                    var newApp = LogIn(config, p.Value.Credential);
+                    // TODO: Log login failure error.
+
+                    if (newApp != null)
+                    {
+                        // Only update if we have a new application.
+                        Profiles[p.Key] = new Profile
+                        {
+                            Application = newApp,
+                            Credential = p.Value.Credential,
+                        };
+
+                        if (currApp != null)
+                        {
+                            /*
+                             * Wait to allow requests using the old application
+                             * to complete before releasing it.
+                             */
+                            System.Threading.Thread.Sleep(10000);
```

Comment:
What happens if requests take more than 10 seconds to complete? Is there a way to wait for them to complete?

Reply:
It's not the whole request we have to wait for. It's the time to get the sessionid and create a new application. Once the new application is created then this session can go away and the created one ca do it's work. So the only time that span we have to cover is getting the sessionid attribute and then calling OnBase login, we don't have to wait for the request to complete.